### ISSUE TITLE
Numerical Drift between `nn.DataParallel` and `Accelerate`

### ISSUE THREAD

**Issue by kheyer** [huggingface/accelerate#2015]
### System Info

```Shell
- `Accelerate` version: 0.23.0
- Platform: Linux-5.15.0-84-generic-x86_64-with-glibc2.31
- Python version: 3.9.18
- Numpy version: 1.26.0
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 125.65 GB
- GPU type: NVIDIA GeForce RTX 3090
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: MULTI_GPU
        - mixed_precision: fp16
        - use_cpu: False
        - debug: False
        - num_processes: 2
        - machine_rank: 0
        - num_machines: 1
        - rdzv_backend: static
        - same_network: False
        - main_training_function: main
        - downcast_bf16: False
        - tpu_use_cluster: False
        - tpu_use_sudo: False
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

I'm noticing numerical drift when computing embeddings between parallel processing with `nn.DataParallel` and `Accelerate`

Pytorch `nn.DataParallel` script:

```python
from transformers import RobertaModel, RobertaTokenizerFast, DataCollatorWithPadding
from datasets import load_dataset, disable_caching
disable_caching()
import torch

def tokenize(examples):
    return tokenizer(examples['text'], truncation=True, padding='max_length', max_length=512)

def embed(inputs):
    inputs = {k: inputs[k] for k in ['input_ids', 'attention_mask']}
    inputs = collator(inputs)
        
    with torch.no_grad():
        outputs = model(**inputs, output_hidden_states=True)
        full_embeddings = outputs[-1][-1]
        mask = inputs['attention_mask']
        
        mask = mask.to(full_embeddings.device)

        mean_embeddings = ((full_embeddings * mask.unsqueeze(-1)).sum(1) / mask.sum(-1).unsqueeze(-1))
        
    return {'embedding' : mean_embeddings}

model_name = "roberta-base"
tokenizer = RobertaTokenizerFast.from_pretrained(model_name)
collator = DataCollatorWithPadding(tokenizer, padding=True, return_tensors='pt')

dataset = load_dataset('imdb')['train']
dataset = dataset.map(tokenize, batched=True, num_proc=64)

model = RobertaModel.from_pretrained(model_name).eval()
model = torch.nn.DataParallel(model, device_ids=[0, 1]).cuda()

dataset = dataset.map(embed, batched=True, batch_size=64)
dataset = dataset.remove_columns(['text', 'label', 'input_ids', 'attention_mask'])
dataset.save_to_disk('emb1.hf')
```

Accelerate script:

```python
from tqdm import tqdm
import datasets
from datasets import disable_caching, load_dataset
disable_caching()
from accelerate import Accelerator

import torch
from torch.utils.data import DataLoader

from transformers import RobertaTokenizerFast, RobertaForMaskedLM, DataCollatorWithPadding

model_name = "roberta-base"
embed_batch_size = 64

tokenizer = RobertaTokenizerFast.from_pretrained(model_name)
collator = DataCollatorWithPadding(tokenizer, padding=True, return_tensors='pt')

def tokenize(examples):
    return tokenizer(examples['text'], truncation=True, padding='max_length', max_length=512)

def main():
    embeddings = []
    accelerator = Accelerator()
    
    with accelerator.main_process_first():
        model = RobertaForMaskedLM.from_pretrained(model_name).eval()
        dataset = load_dataset('imdb')['train']
        dataset = dataset.map(tokenize, batched=True, num_proc=64)
        dataset = dataset.remove_columns(['text', 'label'])
    
    dataloader = DataLoader(dataset, batch_size=embed_batch_size, collate_fn=collator, drop_last=False)
    
    model, dataloader = accelerator.prepare(model, dataloader)
    
    with torch.no_grad():
        for batch in tqdm(dataloader):
            unwrap_model = accelerator.unwrap_model(model)
            outputs = unwrap_model(**batch, output_hidden_states=True)

            full_embeddings = outputs[-1][-1]
            mask = batch['attention_mask']
            mask = mask.to(full_embeddings.device)
            mean_embeddings = ((full_embeddings * mask.unsqueeze(-1)).sum(1) / mask.sum(-1).unsqueeze(-1))
            embeddings += accelerator.gather_for_metrics(mean_embeddings).detach().cpu()
            
    if accelerator.is_main_process:
        embedded_dataset = datasets.Dataset.from_dict({'embedding':embeddings})
        embedded_dataset.save_to_disk('emb2.hf')
                        
    return embeddings
    
if __name__ == "__main__":
    main()
```

Compare embeddings:

```python
import datasets

dataset1 = datasets.load_from_disk('emb1.hf')
dataset2 = datasets.load_from_disk('emb2.hf')

dataset1 = dataset1.with_format('pt')
dataset2 = dataset2.with_format('pt')

deltas = (dataset1['embedding'] - dataset2['embedding']).abs().mean(-1)

deltas.mean()
> tensor(5.2822e-05)

deltas.min()
> tensor(2.4060e-05)
```

### Expected behavior

Expected behavior is embeddings from both methods would be the same.

For each method, we use the same model in eval mode. Both methods use fp32 precision. (maybe accelerate is doing mixed precision behind the scenes in a way I'm not aware of? but the computed embeddings have full precision).

Is there an expected source of numerical drift that would explain this that I am not aware of? Or is the discrepancy unexpected?

**Comment by muellerzr** [comment]
You have mixed precision set to fp16 through your config. This is *automatic mixed precision* which is something accelerate advertises. So yes, your model passed through prepare() is in fp16 with the outputs converted to fp32

**Comment by kheyer** [comment]
Is the mixed precision implemented differently from pytorch? Using `torch.autocast`  and FP16 inference and converting the outputs to FP32 I can recover a small portion of embeddings, but most still show a small difference.

```python
from transformers import RobertaModel, RobertaTokenizerFast, DataCollatorWithPadding
from datasets import load_dataset, disable_caching
disable_caching()
import torch

def tokenize(examples):
    return tokenizer(examples['text'], truncation=True, padding='max_length', max_length=512)

def embed(inputs):
    inputs = {k: inputs[k] for k in ['input_ids', 'attention_mask']}
    inputs = collator(inputs)
        
    with torch.autocast(device_type='cuda', dtype=torch.float16):
        with torch.no_grad():
            outputs = model(**inputs, output_hidden_states=True)
            full_embeddings = outputs[-1][-1]
            
    mask = inputs['attention_mask']
    mask = mask.to(full_embeddings.device)

    mean_embeddings = ((full_embeddings * mask.unsqueeze(-1)).sum(1) / mask.sum(-1).unsqueeze(-1))
        
    return {'embedding' : mean_embeddings.float()}

model_name = "roberta-base"
tokenizer = RobertaTokenizerFast.from_pretrained(model_name)
collator = DataCollatorWithPadding(tokenizer, padding=True, return_tensors='pt')

dataset = load_dataset('imdb')['train']
dataset = dataset.map(tokenize, batched=True, num_proc=64)

model = RobertaModel.from_pretrained(model_name).eval()
model = torch.nn.DataParallel(model, device_ids=[0, 1]).cuda()

dataset = dataset.map(embed, batched=True, batch_size=64)
dataset = dataset.remove_columns(['text', 'label', 'input_ids', 'attention_mask'])
dataset.save_to_disk('emb1.hf')
```

```python
import datasets

dataset1 = datasets.load_from_disk('emb1.hf')
dataset2 = datasets.load_from_disk('emb2.hf')

dataset1 = dataset1.with_format('pt')
dataset2 = dataset2.with_format('pt')

deltas = (dataset1['embedding'] - dataset2['embedding']).abs().mean(-1)

deltas.mean()
> (tensor(5.1000e-05)

deltas.min()
> tensor(0.)

(deltas==0).float().mean()
> tensor(0.0016))
```

**Comment by muellerzr** [comment]
One thing helpful to me that is unclear:

Are you using `accelerate launch` with both scripts? or `accelerate launch` with one and `python` with the other, etc

**Comment by muellerzr** [comment]
I have a few questions here:

1. How are you running these scripts? Via `python`, `accelerate launch`, or `torchrun` for each
2. Why are we then *unwrapping* the model and removing the `nn.DataParallel` if you're launching on two GPUs?

More context on what we're fully doing here will help me understand the problem in full
